File: modules/publix/app/session/batch/akka/actors/services/BatchActionMsgBuilder.java
package session.batch.akka.actors.services;

import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Singleton;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.google.common.base.Strings;

import daos.common.BatchDao;
import models.common.Batch;
import play.Logger;
import play.Logger.ALogger;
import play.db.jpa.JPAApi;
import play.libs.Json;
import session.batch.akka.protocol.BatchDispatcherProtocol.BatchActionMsg;
import session.batch.akka.protocol.BatchDispatcherProtocol.BatchActionMsg.BatchAction;
import session.batch.akka.protocol.BatchDispatcherProtocol.BatchActionMsg.TellWhom;

/**
 * Utility class that builds BatchActionMsgs. So it mostly handles the JSON node
 * creation.
 * 
 * @author Kristian Lange (2017)
 */
@Singleton
public class BatchActionMsgBuilder {

	private static final ALogger LOGGER = Logger
			.of(BatchActionMsgBuilder.class);

	private final JPAApi jpa;

	private final BatchDao batchDao;

	@Inject
	public BatchActionMsgBuilder(JPAApi jpa, BatchDao batchDao) {
		this.jpa = jpa;
		this.batchDao = batchDao;
	}

	/**
	 * Builds a simple BatchActionMsg with the action and the session version
	 */
	public BatchActionMsg buildSimple(Batch batch, BatchAction action,
			TellWhom tellWhom) {
		LOGGER.debug(".buildSimple: batchId {}", batch.getId());
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, action.toString());
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

	/**
	 * Creates a simple BatchActionMsg with an error message
	 */
	public BatchActionMsg buildError(String errorMsg, TellWhom tellWhom) {
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, BatchAction.ERROR.toString());
		objectNode.put(BatchActionMsg.ERROR_MSG, errorMsg);
		return new BatchActionMsg(objectNode, tellWhom);
	}

	/**
	 * Builds a BatchActionMessage with the batch session patch and version
	 */
	public BatchActionMsg buildSessionPatch(Batch batch,
			JsonNode batchSessionPatchNode, TellWhom tellWhom) {
		LOGGER.debug(".buildSessionPatch: batchId {}", batch.getId());
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, BatchAction.SESSION.toString());
		objectNode.set(BatchActionMsg.BATCH_SESSION_PATCHES,
				batchSessionPatchNode);
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

	/**
	 * Builds a BatchActionMsg with the current batch session data and version
	 */
	public BatchActionMsg buildSessionData(long batchId, BatchAction action,
			TellWhom tellWhom) {
		LOGGER.debug(".buildSessionData: batchId {}, action {}, tellWhom {}",
				batchId, action, tellWhom.name());
		return jpa.withTransaction(() -> {
			Batch batch = batchDao.findById(batchId);
			if (batch != null) {
				return buildSessionActionMsg(batch, action, tellWhom);
			} else {
				String errorMsg = "Couldn't find batch with ID " + batchId
						+ " in database.";
				return buildError(errorMsg, TellWhom.SENDER_ONLY);
			}
		});
	}

	private BatchActionMsg buildSessionActionMsg(Batch batch,
			BatchAction action, TellWhom tellWhom) {
		ObjectNode objectNode = Json.mapper().createObjectNode();
		objectNode.put(BatchActionMsg.ACTION, action.toString());
		try {
			if (Strings.isNullOrEmpty(batch.getBatchSessionData())) {
				objectNode.set(BatchActionMsg.BATCH_SESSION_DATA,
						Json.mapper().createObjectNode());
			} else {
				objectNode.set(BatchActionMsg.BATCH_SESSION_DATA,
						Json.mapper().readTree(batch.getBatchSessionData()));
			}
		} catch (IOException e) {
			LOGGER.error(
					".buildSessionActionMsg: invalid session data in DB -"
							+ " batchId {}, batchSessionVersion {},"
							+ " batchSessionData {}, error: {}",
					batch.getId(), batch.getBatchSessionVersion(),
					batch.getBatchSessionData(), e.getMessage());
			objectNode.set(BatchActionMsg.BATCH_SESSION_DATA,
					Json.mapper().createObjectNode());
		}
		objectNode.put(BatchActionMsg.BATCH_SESSION_VERSION,
				batch.getBatchSessionVersion());
		return new BatchActionMsg(objectNode, tellWhom);
	}

}

File: modules/publix/app/session/batch/BatchActionMsgBuilder.scala
package session.batch

import java.io.IOException
import javax.inject.{Inject, Singleton}

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.base.Strings
import daos.common.BatchDao
import models.common.Batch
import play.api.Logger
import play.db.jpa.JPAApi
import play.libs.Json
import session.batch.BatchDispatcher.BatchAction.BatchAction
import session.batch.BatchDispatcher.TellWhom.TellWhom
import session.batch.BatchDispatcher.{BatchAction, BatchActionJsonKey, BatchActionMsg, TellWhom}

import scala.compat.java8.FunctionConverters.asJavaSupplier

/**
  * Utility class that builds BatchActionMsgs. So it mostly handles the JSON node
  * creation.
  *
  * @author Kristian Lange (2017)
  */
@Singleton
class BatchActionMsgBuilder @Inject()(jpa: JPAApi, batchDao: BatchDao) {

  private val logger: Logger = Logger(this.getClass)

  /**
    * Creates a simple BatchActionMsg with an error message
    */
  def buildError(errorMsg: String, tellWhom: TellWhom) = {
    val objectNode = Json.mapper.createObjectNode
    objectNode.put(BatchActionJsonKey.Action.toString, BatchAction.Error.toString)
    objectNode.put(BatchActionJsonKey.ErrorMsg.toString, errorMsg)
    BatchActionMsg(objectNode, tellWhom)
  }

  /**
    * Builds a simple BatchActionMsg with the action and the session version
    */
  def buildSimple(batch: Batch, action: BatchAction, tellWhom: TellWhom) = {
    logger.debug(s".buildSimple: batchId ${batch.getId}")
    val objectNode = Json.mapper.createObjectNode
    objectNode.put(BatchActionJsonKey.Action.toString, action.toString)
    objectNode.put(BatchActionJsonKey.SessionVersion.toString, batch.getBatchSessionVersion)
    BatchActionMsg(objectNode, tellWhom)
  }

  /**
    * Builds a BatchActionMessage with the batch session patch and version
    */
  def buildSessionPatch(batch: Batch, sessionPatchNode: JsonNode, tellWhom: TellWhom) = {
    logger.debug(s".buildSessionPatch: batchId ${batch.getId}")
    val objectNode = Json.mapper.createObjectNode
    objectNode.put(BatchActionJsonKey.Action.toString, BatchAction.Session.toString)
    objectNode.set(BatchActionJsonKey.SessionPatches.toString, sessionPatchNode)
    objectNode.put(BatchActionJsonKey.SessionVersion.toString, batch.getBatchSessionVersion)
    BatchActionMsg(objectNode, tellWhom)
  }

  /**
    * Builds a BatchActionMsg with the current batch session data and version
    */
  def buildSessionData(batchId: Long, action: BatchAction, tellWhom: TellWhom) = {
    jpa.withTransaction(asJavaSupplier(() => {
      logger.debug(s".buildSessionData: batchId $batchId, action $action, tellWhom ${tellWhom.toString}")
      val batch = batchDao.findById(batchId)
      if (batch != null) buildSessionActionMsg(batch, action, tellWhom)
      else buildError("Couldn't find batch with ID " + batchId + " in database.", TellWhom.SenderOnly)
    }))
  }

  private def buildSessionActionMsg(batch: Batch, action: BatchAction, tellWhom: TellWhom) = {
    val objectNode = Json.mapper.createObjectNode
    objectNode.put(BatchActionJsonKey.Action.toString, action.toString)
    try
        if (Strings.isNullOrEmpty(batch.getBatchSessionData))
          objectNode.set(BatchActionJsonKey.SessionData.toString,
            Json.mapper.createObjectNode)
        else objectNode.set(BatchActionJsonKey.SessionData.toString,
          Json.mapper.readTree(batch.getBatchSessionData))
    catch {
      case e: IOException =>
        logger.error(s".buildSessionActionMsg: invalid session data in DB - batchId ${batch.getId}, batchSessionVersion ${batch.getBatchSessionVersion}, batchSessionData ${batch.getBatchSessionData}, error: ${e.getMessage}")
        objectNode.set(BatchActionJsonKey.SessionData.toString, Json.mapper.createObjectNode)
    }
    objectNode.put(BatchActionJsonKey.SessionVersion.toString, batch.getBatchSessionVersion)
    BatchActionMsg(objectNode, tellWhom)
  }

}
